File: 3020-find-the-maximum-number-of-elements-in-subset/3020-find-the-maximum-number-of-elements-in-subset.py
from collections import Counter

class Solution:
    def maximumLength(self, nums: list[int]) -> int:
        freq = Counter(nums)
        cnt = freq[1]
        if cnt % 2 == 0 and cnt > 0:
            cnt -= 1
            
        ans = 0
        unq_nums = sorted([key for key in freq.keys() if key != 1])
        
        for num in unq_nums:
            current_val = num
            tempcnt = 0
            flag = False
            
            while True:
                if current_val > 10**9 or current_val not in freq:
                    flag = True
                    break
                if freq[current_val] < 2:
                    tempcnt += 1
                    break
                
                tempcnt += 2
                current_val *= current_val
            if flag and (tempcnt % 2) == 0 and tempcnt > 0:
                tempcnt -= 1
                
            ans = max(ans, tempcnt)
            
        return max(ans, cnt)

File: 3020-find-the-maximum-number-of-elements-in-subset/3020-find-the-maximum-number-of-elements-in-subset.cpp
#include <vector>
#include <array>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution {
public:
    static constexpr int len = 20;
    static constexpr std::array<int, len + 1> powerof2 = []() {
        std::array<int, len + 1> arr{};
        for (int i = 0; i <= len; ++i) {
            arr[i] = 1 << i;
        }
        return arr;
    }();
    
    long long binpow(long long a, long long b) {
        long long res = 1;
        while (b > 0) {
            if (b & 1) {
                res = (1LL * res * a);
            }
            b >>= 1;
            if (b > 0) { 
                a = (1LL * a * a);
            }
        }
        return res;
    }
    
    int maximumLength(vector<int>& nums) {
        sort(begin(nums), end(nums));
        vector<int> temp;
        int cnt = 0;
        unordered_map<long long, int> freq;
        
        for(const auto& itr: nums){
            if(itr != 1){
                temp.emplace_back(itr);
                freq[itr]++;
            } else {
                cnt++;
            }
        }
        vector<int> unq_nums = temp;
        int ans = 0;
        unq_nums.erase(unique(unq_nums.begin(), unq_nums.end()), unq_nums.end());
        for(int index = 0; index < unq_nums.size(); index++){
            int num = unq_nums[index];
            int tempcnt = 0;
            int e = 0;
            bool flag = false;
            while (true) {
                long long current_val = binpow(num, powerof2[e]);
                if (current_val > 1e9 || freq.find(current_val) == freq.end()) {
                    flag = true;
                    break;
                }
                if (freq[current_val] < 2) {
                    tempcnt++;
                    break; 
                }
                
                tempcnt += 2;
                e++;
            }
            
            if(flag == true && (tempcnt % 2) == 0 && tempcnt > 0) tempcnt--;
            ans = max(ans, tempcnt);
        }
        if (cnt % 2 == 0 && cnt > 0) {
            cnt--;
        }
        
        return max(ans, cnt);
    }
};